### DISCUSSION TITLE
JCB35N2 controller protocol

### DISCUSSION THREAD

**Discussion by auchter** [phord/Jarvis#4] [General]
I started playing around with my Jarvis desk today, which uses the simple two-button handset with a JCB35N2 controller (which appears to be different than the FullyCB2C-A controller documented in the README).

It appears that the pinout is the same for this controller (at least for HS0, HS1, Tx, Rx, Gnd, +5V). However, the serial protocol looks to be different.

When moving the desk up/down (by shorting HS0/HS1 to ground), the JCB35N2 controller starts repeatedly sending `0x01 0x01 0x01 <h>`, where h is a byte proportional to the height of the desk. This continues for a few seconds after the desk has stopped moving, and the final message sent by the controller is `0x01 0x05 0x01 0xAA`.

I've tried mimicking the startup sequence that was documented in the README (0x00, 230ms break, 0x00, sending `0xF1 0xF1 0x29 0x00 0x29 0x7E` repeatedly, etc.) but that doesn't seem to change anything.

I'm curious if the messages I'm seeing are the "other UART control protocol" that @phord has seen with his controller, or if this controller is using an altogether different protocol.

**Comment by phord** [comment]
This does sound like the other protocol I sometimes saw after some idle time.  When it happens to me, the display appears to stop working and my daughtercard can no longer read the protocol correctly.  I examined it a few times on the protocol analyzer and saw a sequence of four bytes similar to yours.  I decoded the output like this:

    0x01 0x01 0x04 0xEE
              ==== ====
               HEIGHT

I captured this mode occurring once on the analyzer.    From my notes:

     Pressed the UP button and saw 0x81 0xF5 (with framing errors) on Tx. This was followed by 0x01 0x01 0x01 0xEE.
     This 4-byte code repeated for 11 seconds.  
     Finally it sent 0x01 0x05.  (Display off?)

I didn't see the `0x01 0xAA` in my capture.  I wonder if those are just "empty registers" though.  0xAA is an interesting pattern of alternating zeros and ones.  `0b10101010`  Maybe just a coincidence.

**Comment by kennyyork** [comment]
@auchter I'm seeing similar behavior with mine, but consistently.  I have a simple Arduino sketch to just dump out any serial data on the line and as soon as I power up the desk I get a stream of `0x01 0x01 0x01 0xB0` for a few seconds then it stop.  Moving the desk up and down seems to generate a changing stream of values that I assume correlate to the height.

Did you ever make any progress?

**Reply by kennyyork** [reply]
Interesting.  My handset was in the repeated 0x01 mode until I issued a 230ms break followed by a WAKE command.  I wonder if some logic in the desk is looking for that signal to swap into the "advanced" mode?

Maybe while testing the controller crashes, rebooting in "simple" mode until you send a BREAK/WAKE?

**Reply by kennyyork** [reply]
And now I can't get it to go back in to "advanced" mode.

**Reply by phord** [reply]
@kennyyork What model controller and handset do you have?

**Reply by kennyyork** [reply]
All I have is the basic Up/Down handset.  The controller under the desk is FULLYCB2C-A.

I'm not sure what I did to make it happy and start sending the framed packets like you documented, but I've only done that twice now.   Right now when I start up and write a 0, its sends me a stream of simple position data for about 8-10 seconds. 

After that if I issue a break it sends the simple position data again.

**Reply by phord** [reply]
It's exciting that you got it to send the more elaborate stream at all. That means it is in there. We just need to figure out how activate it.

On the other hand, maybe the simple protocol is enough for your needs. If it sends height information, you can write the display and memory function yourself in software. 

When I get a chance I'll try to experiment more with the simple protocol. Probably in a couple of weeks.

**Reply by kennyyork** [reply]
That would be great since I don't have a memory handset to play with.

My real goal is to program my sit height to slot 1 and my stand height to slot 2.  Then I should be able to just activate those two positions via GPIO pins, correct?   Once I have that setup I already have a RaspberryPi running as a HomeKit accessory and I'll wire that into the on/off state.  Then I can control the desk with Siri and Shortcuts

**Reply by phord** [reply]
Touch the "Up" gpio for 50ms to get the unit to send the height. If the height is correct, you're done. If it's too low, hold Up. If it's too high, hold Down. 

Release when it's correct or "close enough".

**Comment by OkhammahkO** [comment]
I think you guys might be interested in this project. 

We've cracked a lot of Jiecang based desks.

https://community.home-assistant.io/t/desky-standing-desk-esphome-works-with-desky-uplift-jiecang-assmann-others/383790?u=mahko_mahko

**Reply by OkhammahkO** [reply]
Interestingly on my Jiecang based desk, you get the different protocols on the rj45 port (for handset) and the rj12 port (for Bluetooth dongle).

Height updates on the rj12 port are less frequent (perhaps to limit traffic over Bluetooth).

**Reply by phord** [reply]
That's interesting.  Have you looked at my protocol notes to see if we both found all the same commands?

**Reply by OkhammahkO** [reply]
Not yet, but I'm planning on taking a close look soonish....

It looks like you've cracked quite a lot....

**Reply by OkhammahkO** [reply]
Hi @phord.

I'm finally getting around to seeing if the commands you have discovered work for the ESPHome Desky project. 
I'm pretty green with UART.
From your write-up, I interpret a message to set the memory #3 slot to current height as being as follows:

```
#Code snippet in ESPHome yaml format:

button:
  - platform: template
    name: Set memory 3
    id: set_memory_3
    on_press:
      then:
        - uart.write: [0xF1, 0xF1, 0x25, 0x00, 0x25, 0x7E]
```
Does that look right?
It didn't seem to work though. Do you have to "put it into memory mode first" via the special HS3 command before issuing this command? 

I don't have that particular wire hooked up a the moment and it's a bit of a hassle to pull apart my diy handset so I thought I'd just ask.

Cheers.

**Reply by phord** [reply]
That does look right to me.

**Comment by Karuption** [comment]
Adding to this, I was asked to capture the start up for the JCB35N2 controller and the 1 1 <height> protocol. There are .sal files for the captures in the .zip. I am not sure what the best way to share this would be.

[StartUp.zip](https://github.com/phord/Jarvis/files/10937425/StartUp.zip)

**Comment by phord** [comment]
Thanks for the captures, @Karuption .  That's a lot of signaling going on in your handset.  A lot more complicated than the Fully protocol.  Can you describe the conditions for the two captures?  
I think the StartupNHC is where you powered on the desk with the handset disconnected first; then you connected the handset.
Is StartupWHC captured from powering on the desk with the handset already connected?

**Reply by Karuption** [reply]
Yes, that's correct. The channels are named correct according to your pinouts.   I am not really sure how to take all the volatility as it seems that none of that is UART. I would like to say that it's because of the cheap logic analyser, but i don't have anything to confirm it with. 

I labeled them start up with hand controller and start up no hand controller.

**Comment by Karuption** [comment]
@phord I finally got around to probing the RJ11 jack on my desk. Interestingly, it seems that it is running The xF2xF2 protocol for that port. Here is a snippet of lowering my desk from the console output on Logic.

`\xF2\xF2\x01\x03\x01\x05\x0F\x19~\xF2\xF2\x01\x03\x01\x04\x0F\x18~\xF2\xF2\x01\x03\x01\x02\x0F\x16~\xF2\xF2\x01\x03\x01\x01\x0F\x15~\xF2\xF2\x01\x03\0\xFF\x0F\x12~\xF2\xF2\x01\x03\0\xFE\x0F\x11~\xF2\xF2\x01\x03\0\xFD\x0F\x10~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~\xF2\xF2\x01\x03\0\xFC\x0F\x0F~`

**Reply by OkhammahkO** [reply]
It could be the same or similar to this....

https://community.home-assistant.io/t/desky-standing-desk-esphome-works-with-desky-uplift-jiecang-assmann-others/383790/23?u=mahko_mahko

**Reply by Karuption** [reply]
This repo's main focus is about the same controller. I have the "other" protocol controller (that this thread in focused on) but, when looking at the port on the controller, I am seeing the packet based protocol. 

RJ-45 - broadcast based
RJ-12 - packet based

**Comment by rickdgray** [comment]
Sorry, is this project yet usable for 2 button desks? This discussion seems to be in the realm of embedded systems which is slightly outside my area of knowledge so I can't quite make sense of it. I am happy to crimp up an RJ-45 and start experimenting; I would love for this to work on my desk. Any way I can help?

**Reply by Karuption** [reply]
Yes, it should work with any hand unit. There are two major protocols issues over RJ-45 so, it hopefully is just setting up the correct pin out, flashing the controller and figuring out what one of those two it is.
